test/parallel/test-assert-deep.js

```diff
@@ -1346,3 +1361,107 @@ test('Comparing two different WeakSet instances', () => {
   const weakSet2 = new WeakSet();
   assertNotDeepOrStrict(weakSet1, weakSet2);
 });
+
+test('Comparing two arrays nested inside object, with overlapping elements', () => {
+  const actual = { a: { b: [1, 2, 3] } };
+  const expected = { a: { b: [3, 4, 5] } };
+
+  assert.throws(
+    () => assert.deepStrictEqual(actual, expected),
+    {
+      code: 'ERR_ASSERTION',
+      name: 'AssertionError',
+      message: 'Expected values to be strictly deep-equal:\n' +
+        '+ actual - expected\n' +
+        '\n' +
+        '  {\n' +
+        '    a: {\n' +
+        '      b: [\n' +
+        '+       1,\n' +
+        '+       2,\n' +
+        '        3,\n' +
+        '-       4,\n' +
+        '-       5\n' +
+        '      ]\n' +
+        '    }\n' +
+        '  }\n'
+    }
+  );
+});
+
+test('Comparing two arrays nested inside object, with overlapping elements, swapping keys', () => {
+  const actual = { a: { b: [1, 2, 3], c: 2 } };
+  const expected = { a: { b: 1, c: [3, 4, 5] } };
+
+  assert.throws(
+    () => assert.deepStrictEqual(actual, expected),
+    {
+      code: 'ERR_ASSERTION',
+      name: 'AssertionError',
+      message: 'Expected values to be strictly deep-equal:\n' +
+      '+ actual - expected\n' +
+      '\n' +
+      '  {\n' +
+      '    a: {\n' +
+      '+     b: [\n' +
+      '+       1,\n' +
+      '+       2,\n' +
+      '+       3\n' +
+      '-     b: 1,\n' +
+      '-     c: [\n' +
+      '-       3,\n' +
+      '-       4,\n' +
+      '-       5\n' +
+      '      ],\n' +
+      '+     c: 2\n' +
+      '    }\n' +
+      '  }\n'
+    }
+  );
+});
+
+test('Detects differences in deeply nested arrays instead of seeing a new object', () => {
+  const actual = [
+    { a: 1 },
+    2,
+    3,
+    4,
+    { c: [1, 2, 3] },
+  ];
+  const expected = [
+    { a: 1 },
+    2,
+    3,
+    4,
+    { c: [3, 4, 5] },
+  ];
+
+  assert.throws(
+    () => assert.deepStrictEqual(actual, expected),
+    {
+      code: 'ERR_ASSERTION',
+      name: 'AssertionError',
+      message: 'Expected values to be strictly deep-equal:\n' +
+        '+ actual - expected\n' +
+        '... Skipped unchanged lines\n' +
+        '... Skipped added lines\n' +
+        '... Skipped removed lines\n' +
```

Comment:
Differentiating these does not seem like it improves things for the reader.

If `...` is in-between unchanged lines, it's clear that they are all unchanged. The same applies to added or removed lines.

Reply:
Yes, you are right if the colors were not existing :) but whenever they do, you see something similar:

![image](https://github.com/user-attachments/assets/5232b88c-b765-4cf4-96c2-e8ca9d8437b5)

but I am not really super into this either... if you think we should remove the differentiation, I will 😄 just say so and it will be done!

Reply:
the difference between white `...` and blue or red `...` is that with white dots the lines actually didn't change but with blue or red ones the lines actually changed, but they were repeated so many times ( and all in a row!) that it was useless to print all of them.  
Btw, I applied your suggestion, maybe the end user is smart enough to figure out the difference by himself :D

Reply:
I just had another look at this and I would rather not hide any added or removed lines at all. The reason is, that each time we hide something, it becomes more likely to hide information that would map it to the actual input. Imagine an object with multiple objects inside that all look somewhat similar but differ.

Printing too many lines is, however, also tricky and not ideal to read. There is therefore definitely a line limit that would make sense (likely > 1000 lines overall) to trigger this optimization (while it also becomes more likely to confuse the reader). We could also just print `... identical added lines` and `... identical removed lines` on each entry in that case. That works without colors as well.